### PR TITLE
Exclude flaky tests using annotation

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

It is a hassle to add `assumeThat(..)` for excluding flaky tests.
It is also not a good way to exclude a whole class or its subclass.
For convenient sake, we can use JUnit5's `@Tag` annotation and filter in or out
using `includeTags` and `excludeTags`.

Modifications:

- Add `@FlakyTest` annotation to exclude a flaky test method or class
  from the smoke test.
- Mark `GracefulShutdownIntegrationTest` as a flaky test.

Result:

Easy to maintain flaky tests

### CODE REVIEW

File: .github/workflows/actions_build.yml [hunk 1]
```diff
@@ -82,12 +82,11 @@ jobs:
         ${{ matrix.coverage && '-Pcoverage' || '' }} \
         ${{ matrix.leak && '-Pleak' || '' }} \
         -PnoLint \
+        -PflakyTest=off \
```
Comment (minwoox) [inline .github/workflows/actions_build.yml:85]: nit: How about using `true` and `false`?
I'm not strong on this so `on` and `off` are also find to me. 😄

**Review by minwoox — APPROVED** [review]
Thanks!

File: .github/workflows/actions_build.yml (hunk 1 above)
Comment (ikhoon) [reply .github/workflows/actions_build.yml:85]: Either is fine with me. `true` and `false` sounds better.

**Review by anuraaga — COMMENTED** [review]
Any thoughts on using gradle-retry-plugin instead? If enabling build scans, which is a nice idea in general I think, retries get marked out specially too

https://github.com/gradle/test-retry-gradle-plugin

Could be in addition to the tag but not sure if tag is needed if there is auto retry

**Comment by ikhoon** [comment]
> Any thoughts on using gradle-retry-plugin instead? 

That sounds nice. But I am not sure retries really solve our headache flaky problems. We manually keep trying to start an embedded Consul for 40 seconds, but we still easily see some failed tests in Consul module.
https://github.com/line/armeria/blob/0f7a07750f4afa42618c53a66fe08bf30d9f5a70/consul/src/test/java/com/linecorp/armeria/internal/consul/ConsulTestBase.java#L116-L122
In the end, we decided to split flaky tests from the smoke test.

File: gradle/scripts/lib/java.gradle [hunk 2]
```diff
@@ -92,7 +92,18 @@ configure(projectsWithFlags('java')) {
...
+        if (flakyTest == 'true') {
+            useJUnitPlatform {
+                includeTags 'FLAKY_TEST'
+            }
+        } else if (flakyTest == 'false') {
+            useJUnitPlatform {
+                excludeTags 'FLAKY_TEST'
+            }
+        } else {
+            useJUnitPlatform()
```
Comment (trustin) [inline gradle/scripts/lib/java.gradle:105]: How about raising an exception if `flakyTest` is not `null`?

**Review by trustin — APPROVED** [review]

File: gradle/scripts/lib/java.gradle (hunk 2 above)
Comment (ikhoon) [reply gradle/scripts/lib/java.gradle:105]: Sounds nice. 👍

**Comment by trustin** [comment]
@ikhoon Please send a PR to https://github.com/line/gradle-scripts.

File: gradle/scripts/lib/java.gradle [hunk 3]
```diff
@@ -92,7 +92,21 @@ configure(projectsWithFlags('java')) {
...
         // Use JUnit platform.
-        useJUnitPlatform()
+        def flakyTest = rootProject.findProperty('flakyTest')
+        if (flakyTest == 'true') {
+            useJUnitPlatform {
+                includeTags 'FLAKY_TEST'
+            }
+        } else if (flakyTest == 'false') {
+            useJUnitPlatform {
+                excludeTags 'FLAKY_TEST'
```
Comment (trustin) [inline gradle/scripts/lib/java.gradle:102]: Should we use a plural form, e.g. `flakyTests` and `FLAKY_TESTS`?

Comment (ikhoon) [reply gradle/scripts/lib/java.gradle:102]: Initially, I used a singular form for consistency with `@FlakyTest` annotation, but a plural form sounds good.

**Comment by ikhoon** [comment]
> Please send a PR to [line/gradle-scripts](https://github.com/line/gradle-scripts).

Sent a PR for this. https://github.com/line/gradle-scripts/pull/103

File: gradle/scripts/lib/java.gradle [hunk 4]
```diff
@@ -92,7 +92,21 @@ configure(projectsWithFlags('java')) {
...
+            useJUnitPlatform {
+                includeTags 'FLAKY_TESTS'
+            }
+        } else if (flakyTests == 'false') {
+            useJUnitPlatform {
+                excludeTags 'FLAKY_TESTS'
+            }
+        } else {
+            if (flakyTests != null) {
+                throw new IllegalArgumentException("flakyTest: $flakyTests (expected: true, false or null)")
```
Comment (trustin) [inline gradle/scripts/lib/java.gradle:106]:
```suggestion
                throw new IllegalArgumentException("flakyTests: $flakyTests (expected: true, false or null)")
```

**Pull request merged**